### ISSUE TITLE
BASE assumes that a set $use_referential_integrity conf var indicates presence of RI DB functionality.. 

### ISSUE THREAD

**Issue by NathanGibbs3** [NathanGibbs3/BASE#215]
Item | Description
-----|-------------
**File(s):** | 
**Class:** | 
**Function:** | 
**Similar Issues:** | #103 #212
**Depends on Issue(s):** | #214 
**Dependency Type:** | Hard
**Misc. Info.:** | 
---
**Expected Behavior:** 
BASE **ONLY** uses RI functionality when it is in the DB.
**Current Behavior:**
BASE assumes that a set `$use_referential_integrity` conf var indicates presence of RI DB functionality..

**Comment by mesteele** [comment]
I need some clarification on this added and updated feature: On an **existing** or **new** DB if I set BASE to $use_referential_integrity = 1 will BASE add the RI tables to the DB?

**Comment by NathanGibbs3** [comment]
That's the idea, if the DB server can support it, it enables RI when set.
It's also supposed to follow the `$use_referential_integrity` setting and remove the RI constraints from the DB when set to zero.

However, I **WOULD NOT** deploy what is currently in devel into the wild yet.
What is currently in devel is a steaming pile of 💩 that only works as intended on MariaDB.

Thank God I built Unit Tests around this, when it hit travis-CI, it went off the rails and is a train wreck.

We are currently having issues with the RI structure verification code on PostgreSQL and RI structure teardown on MySQL which does things just a "little bit" differently than MariaDB.

The PostgreSQL issue is the one I'm working on right now, because on PostgreSQL, enabling RI will crash BASE with this feature. 💩 🔥 🧯 🚒 😄

**Comment by mesteele** [comment]
I'm confused, but that doesn't take much :) Would you feel comfortable putting out the one listed in the "Master" release to the wild or should the last version released by secureideas be used?

I'm assuming that the Master is the last PHP7 compatible version?

**Comment by NathanGibbs3** [comment]
> I'm confused, but that doesn't take much :) Would you feel comfortable putting out the one listed in the "Master" release to the wild or should the last version released by secureideas be used?
> 

The code in master is OK.

The code in devel is all the work around getting RI working on MySQL/ MariaDB, and making sure that even if the end user sets `$use_referential_integrity` to 1 on a DB setup that can't support it, the chance of DB corruption is minimized.

> I'm assuming that the Master is the last PHP7 compatible version?

There are no plans to drop PHP 7x compatibility.
That is part of why I build Unit Tests around what I am working on, to ensure where possible that BASE works on older PHP's as well.

**Comment by NathanGibbs3** [comment]
Good news, the last commit to devel passed CI checks.
 Use it in the wild if you want.

I'll go over the RI stuff again, sorry about the confusion earlier.

Classic BASE 'secureideas' behavior.
- `$use_referential_integrity = 0`
  - BASE deletes alerts from event table.
  - BASE deletes alerts from other tables too.
- `$use_referential_integrity = 1`
  - BASE deletes alerts from event table.
  - THE DB server deletes alerts from other tables, if RI has been setup and the DB server even supports RI.
  - BASE assumes that the DB server took care of it.
 
New BASE behavior.
- `$use_referential_integrity = 0` ( Same as classic at the moment. )
  - BASE deletes alerts from event table.
  - BASE deletes alerts from other tables too.
- `$use_referential_integrity = 1`
  - BASE deletes alerts from event table.
  - THE DB server deletes alerts from other tables, if RI has been setup.
  - When BASE connects to the DB it does the following checks.
    - Server Check: Can DB Server ( Version / Features ) support RI?
      - If Yes, DB Check: ( Is RI already setup in the DB? Is RI structure valid? )
        - If No,
          - Remove parts of invalid RI structure if they exists.
          - Create new RI structure.
        - If Yes, Carry on, nothing more to do concerning RI.
      - If No, DB Server cannot support RI, act like `$use_referential_integrity = 0`

---
Hope that clarifies things for you. :smile:
BTW, great feature request. 🎉

**Comment by mesteele** [comment]
> There are no plans to drop PHP 7x compatibility.

Sorry, the Master has no PHP 8 code included, or does it?

Is the Master what we call stable?

**Comment by NathanGibbs3** [comment]
> > There are no plans to drop PHP 7x compatibility.
> 
> Sorry, the Master has no PHP 8 code included, or does it?
> 

As far as I know master will work with PHP 5x or earlier through PHP 8x.
For instance, the bugs that you found relating to issue #149 have been fixed and those fixes merged into master and tested successfully in travis-CI on PHP 5.2x - 8.1x.

When a PHP forward compatibility bug, ( PHP 7x or 8x breakage ), gets fixed, I try to find a way of doing it that does not break backwards compatibility ( PHP 5x or earlier ).

> Is the Master what we call stable?

It is more stable than devel, although at times mistakes get merged into master, eg. #194
